cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go

```diff
@@ -19,37 +19,79 @@ package uploadconfig
 import (
 	"testing"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 )
 
 func TestUploadConfiguration(t *testing.T) {
 	tests := []struct {
-		name    string
-		cfg     *kubeadmapi.MasterConfiguration
-		wantErr bool
+		name           string
+		createErr      error
+		updateErr      error
+		idempotentTest bool
+		wantErr        bool
+		verifyResult   bool
 	}{
 		{
-			"basic validation with correct key",
-			&kubeadmapi.MasterConfiguration{
-				KubernetesVersion: "1.7.3",
-			},
-			false,
+			name:         "basic validation with correct key",
+			wantErr:      false,
+			verifyResult: true,
+		},
+		{
+			name:           "update existing should report no error",
```

Comment:
I'm not sure if this test is actually exercising the codepath we want it to. Shouldn't we be setting `createErr` to be an `AlreadyExistsError` so it will trigger calling `Update` in the first place? Based on my reading of the test code, it looks like the original `Create` is just going to succeed an we won't trigger the `Update` case.

Reply:
We could do that. But I choose to call UpdateConfiguration twice. Isn't it better to simulate the scenario. The first call to UpdateConfiguration will create config map, the second call to UpdateConfiguration will update configmap

Reply:
Ah, sorry, then I must have not followed the test code correctly. If the test client we use actually keeps track of created objects and returns an `AlreadyExistsError` the second time, then that's awesome.